pkg/statuschecker/ready_checker.go

```diff
@@ -0,0 +1,166 @@
+	ObjectStatusTypeUnknown    ObjectStatusType = "Unknown"
+	ObjectStatusTypeInProgress ObjectStatusType = "InProgress"
+	ObjectStatusTypeError      ObjectStatusType = "Error"
+)
+
+type ObjectStatusResult interface {
+	StatusType() ObjectStatusType
+}
+
+type ObjectStatusReady struct{}
+type ObjectStatusUnknown struct {
+	Details string
+}
+type ObjectStatusInProgress struct {
+	Message string
+}
+type ObjectStatusError struct {
+	UserError      bool
+	RetriableError bool
+	Error          error
+}
+
+func (o ObjectStatusReady) StatusType() ObjectStatusType {
+	return ObjectStatusTypeReady
+}
+func (o ObjectStatusUnknown) StatusType() ObjectStatusType {
+	return ObjectStatusTypeUnknown
+}
+func (o ObjectStatusInProgress) StatusType() ObjectStatusType {
+	return ObjectStatusTypeInProgress
+}
+func (o ObjectStatusError) StatusType() ObjectStatusType {
+	return ObjectStatusTypeError
+}
+
+// ObjectStatusChecker checks object's status.
+// Function is responsible for handling different versions of objects by itself.
+type ObjectStatusChecker func(runtime.Object) (ObjectStatusResult, error)
+
+// CRDStore gets a CRD definition for a Group and Kind of the resource (CRD instance).
+// Returns nil if CRD definition was not found.
+type CRDStore interface {
+	Get(resource schema.GroupKind) (*apiext_v1b1.CustomResourceDefinition, error)
+}
+
+type Interface interface {
+	CheckStatus(*unstructured.Unstructured) (r ObjectStatusResult, e error)
+}
+
+type Checker struct {
+	Store      CRDStore
+	KnownTypes map[schema.GroupKind]ObjectStatusChecker
+}
+
+func New(store CRDStore, kts ...map[schema.GroupKind]ObjectStatusChecker) (*Checker, error) {
+	kt := make(map[schema.GroupKind]ObjectStatusChecker)
+	for _, knownTypes := range kts {
+		for knownGK, f := range knownTypes {
+			if kt[knownGK] != nil {
+				return nil, errors.Errorf("GroupKind specified more than once: %s", knownGK)
+			}
+			kt[knownGK] = f
+		}
+	}
+	return &Checker{
+		Store:      store,
+		KnownTypes: kt,
+	}, nil
+}
+
+func (c *Checker) CheckStatus(obj *unstructured.Unstructured) (ObjectStatusResult, error) {
+	gvk := obj.GroupVersionKind()
+	gk := gvk.GroupKind()
+
+	if gk.Kind == "" || gvk.Version == "" { // Group can be empty e.g. built-in objects like ConfigMap
+		return ObjectStatusUnknown{}, errors.Errorf("object has empty kind/version: %s", gvk)
+	}
+
+	// 1. Check if it is a known built-in resource
+	if isObjectReady, ok := c.KnownTypes[gk]; ok {
+		return isObjectReady(obj)
+	}
+
+	// 2. Check if it is a CRD with path/value annotation
+	crd, err := c.crdWithPathValueAnnotation(gk)
+	if err != nil {
+		return nil, err
+	}
+	if crd != nil {
+		return c.checkPathValue(crd, obj)
+	}
+
+	// 3. Check if it is a CRD with Kind/GroupVersion annotation
+	crd, err = c.crdWithKindGroupVersionAnnotation(gk)
+	if err != nil {
+		return nil, err
+	}
+	if crd != nil {
+		return c.checkForInstance(crd, obj)
+	}
+
+	return ObjectStatusInProgress{}, nil
```

Comment:
I think for unknown objects we should either return an error or ready. in progress does not make any sense. I know this is how it works right now, but we should fix that. Would be great as a follow up if you have time.